internal/integration/api/serviceaccount.go

```diff
@@ -0,0 +1,288 @@
+	})
+
+	suite.Assert().NoError(err)
+}
+
+func (suite *ServiceAccountSuite) waitForSecret(ns, name string) (*corev1.Secret, error) {
+	var (
+		secret *corev1.Secret
+		err    error
+	)
+
+	err = retry.Constant(1*time.Minute).RetryWithContext(suite.ctx, func(ctx context.Context) error {
+		secret, err = suite.Clientset.CoreV1().Secrets(ns).Get(suite.ctx, name, metav1.GetOptions{})
+		if kubeerrors.IsNotFound(err) {
+			return retry.ExpectedError(err)
+		}
+
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+func (suite *ServiceAccountSuite) getCRD() (*unstructured.Unstructured, error) {
+	return suite.DynamicClient.Resource(schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}).Get(suite.ctx, "serviceaccounts.talos.dev", metav1.GetOptions{})
+}
+
+func (suite *ServiceAccountSuite) createServiceAccount(ns string, name string, roles []string) (*unstructured.Unstructured, error) {
+	return suite.DynamicClient.Resource(serviceAccountGVR).Namespace(ns).Create(suite.ctx, &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "talos.dev/v1alpha1",
+			"kind":       "ServiceAccount",
+			"metadata": map[string]interface{}{
+				"name": name,
+			},
+			"spec": map[string]interface{}{
+				"roles": roles,
+			},
+		},
+	}, metav1.CreateOptions{})
+}
+
+// configureAPIAccess configures the API access feature on all control plane nodes.
+func (suite *ServiceAccountSuite) configureAPIAccess(
+	enabled bool,
+	allowedRoles []string,
+	allowedNamespaces []string,
+) error {
+	controlPlaneIPs := suite.DiscoverNodeInternalIPsByType(suite.ctx, machine.TypeControlPlane)
+
+	for _, ip := range controlPlaneIPs {
+		nodeCtx := client.WithNodes(suite.ctx, ip)
+
+		nodeConfig, err := suite.ReadConfigFromNode(nodeCtx)
+		if err != nil {
+			return err
+		}
+
+		nodeConfigRaw, ok := nodeConfig.Raw().(*v1alpha1.Config)
+		if !ok {
+			return fmt.Errorf("unexpected node config type %T", nodeConfig.Raw())
+		}
+
+		accessConfig := v1alpha1.KubernetesTalosAPIAccessConfig{
+			AccessEnabled:                     pointer.To(enabled),
+			AccessAllowedRoles:                allowedRoles,
+			AccessAllowedKubernetesNamespaces: allowedNamespaces,
+		}
+
+		nodeConfigRaw.MachineConfig.MachineFeatures.KubernetesTalosAPIAccessConfig = &accessConfig
+
+		bytes, err := nodeConfigRaw.Bytes()
+		if err != nil {
+			return err
+		}
+
+		_, err = suite.Client.ApplyConfiguration(nodeCtx, &machineapi.ApplyConfigurationRequest{
+			Data: bytes,
+			Mode: machineapi.ApplyConfigurationRequest_NO_REBOOT,
+		})
+
+		if err != nil {
+			return err
+		}
+	}
+
+	if enabled { // wait for CRD to be created
+		return retry.Constant(30*time.Second).RetryWithContext(suite.ctx, func(ctx context.Context) error {
+			_, err := suite.getCRD()
+			if err != nil {
+				return retry.ExpectedError(err)
+			}
+
+			return nil
+		})
+	}
+
+	return nil
+}
+
```

Comment:
shouldn't we have a test to check that the k8s `endpoint` has been created?

Reply:
Good idea, I just added it to the wait block so it will fail the test if the feature is enabled but the service `talos.default` does not appear.